### PR TITLE
Set properties autowired with @required as initialized

### LINKED ISSUE #346: checkUninitializedProperties does not consider @required
When enabling PHPStan's `checkUninitializedProperties`, using `@required` to autowire a public property, or through a public method will still mark the property as uninitialized, although Symfony's autowiring will make sure it isn't.

This is on;
phpstan/phpstan 1.10.13
phpstan/phpstan-symfony 1.3.1

### PR DESCRIPTION
**Pull request by raalderink**
Fixes https://github.com/phpstan/phpstan-symfony/issues/346

### CODE REVIEW

File: src/Symfony/RequiredAutowiringExtension.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+		if (!$property->isPublic()) {
+			return false;
+		}
+
+		if ($property->getDocComment() !== null && $this->isRequiredFromDocComment($property->getDocComment())) {
+			return true;
+		}
+
+		// Check for the attribute version
+		if ($property instanceof PhpPropertyReflection && count($property->getNativeReflection()->getAttributes(Required::class)) > 0) {
```
Comment (ondrejmirtes) [inline src/Symfony/RequiredAutowiringExtension.php]: `Class Symfony\Contracts\Service\Attribute\Required not found.` should be solved by doing `'Symfony\Contracts\Service\Attribute\Required'` as a literal string instead of `::class`.

As for the rest I'm gonna have to think about it and do something in PHPStan itself. Or maybe just ignore them, they're not that severe.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Symfony/RequiredAutowiringExtension.php (hunk 1 above)
Comment (raalderink) [reply src/Symfony/RequiredAutowiringExtension.php]: Most issues are solved now, however using the class as a string, will either error out as it being a `string` and not a `class-string`, but when I explicitly type it as `class-string`, it errors with `PHPDoc tag @var with type class-string is not subtype of native type`. It's kind of fighting both sides this way. Is there perhaps a better solution? Stubs maybe?

Comment (raalderink) [reply src/Symfony/RequiredAutowiringExtension.php]: Okay, it took a little bit of effort figuring out the best way to approach this, but I've fixed it now!

**Comment by raalderink** [comment]
@ondrejmirtes could you re-review this PR please? Thank you :)

File: tests/Symfony/RequiredAutowiringExtensionTest.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+final class RequiredAutowiringExtensionTest extends RuleTestCase
+{
+
+	protected function getRule(): Rule
+	{
+		$container = self::getContainer();
+		$container->getServicesByTag(AdditionalConstructorsExtension::EXTENSION_TAG);
+
+		return new UninitializedPropertyRule(
+			new ConstructorsHelper(
```
Comment (ondrejmirtes) [inline tests/Symfony/RequiredAutowiringExtensionTest.php]: Please fetch the rule with `$container->getByType(UninitializedPropertyRule::class)`. It should be possible. That way it doesn't have to be in the baseline.

File: src/Symfony/RequiredAutowiringExtension.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+			return true;
+		}
+
+		return false;
+	}
+
+	public function getAdditionalConstructors(ClassReflection $classReflection): array
+	{
+		$additionalConstructors = [];
+		/** @var ReflectionClass $nativeReflection */
```
Comment (ondrejmirtes) [inline src/Symfony/RequiredAutowiringExtension.php]: Please remove the inline `@var`.

File: src/Symfony/RequiredAutowiringExtension.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+		if (!$property->isPublic()) {
+			return false;
+		}
+
+		if ($property->getDocComment() !== null && $this->isRequiredFromDocComment($property->getDocComment())) {
+			return true;
+		}
+
+		// Check for the attribute version
+		if (class_exists(Required::class) && $property instanceof PhpPropertyReflection && count($property->getNativeReflection()->getAttributes(Required::class)) > 0) {
```
Comment (ondrejmirtes) [inline src/Symfony/RequiredAutowiringExtension.php]: The class_exists() check here can be removed.

File: tests/Type/Symfony/ExtensionTest.php [hunk 5]
```diff
@@ -14,49 +14,49 @@ class ExtensionTest extends TypeInferenceTestCase
...
 
-		yield from $this->gatherAssertTypes(__DIR__ . '/data/tree_builder.php');
+		yield from self::gatherAssertTypes(__DIR__ . '/data/tree_builder.php');
 
-		yield from $this->gatherAssertTypes(__DIR__ . '/data/ExampleBaseCommand.php');
-		yield from $this->gatherAssertTypes(__DIR__ . '/data/ExampleOptionCommand.php');
-		yield from $this->gatherAssertTypes(__DIR__ . '/data/ExampleOptionLazyCommand.php');
-		yield from $this->gatherAssertTypes(__DIR__ . '/data/kernel_interface.php');
-		yield from $this->gatherAssertTypes(__DIR__ . '/data/request_get_content.php');
+		yield from self::gatherAssertTypes(__DIR__ . '/data/ExampleBaseCommand.php');
```
Comment (ondrejmirtes) [inline tests/Type/Symfony/ExtensionTest.php]: These `$this->` to `self::`  are unrelated to the rest of the PR. Please remove them. If you want, you can submit them in a separate PR. Thanks.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: tests/Symfony/RequiredAutowiringExtensionTest.php (hunk 2 above)
Comment (raalderink) [reply tests/Symfony/RequiredAutowiringExtensionTest.php]: I've done that and removed it from the baseline, however, it still fails on the same thing; 
`Accessing PHPStan\Rules\Properties\UninitializedPropertyRule::class is not covered by backward compatibility promise. The class might change in a minor PHPStan version.`

File: src/Symfony/RequiredAutowiringExtension.php (hunk 3 above)
Comment (raalderink) [reply src/Symfony/RequiredAutowiringExtension.php]: On PHP 7.3 and lower this will throw the following;
`Call to method getMethods() on an unknown class PHPStan\BetterReflection\Reflection\Adapter\ReflectionEnum.`
I'm not quite sure what to do with this, as I'm confused as to why it would do this only on PHP 7.3 and lower

File: composer.json [hunk 6]
```diff
@@ -15,7 +15,8 @@
   "require": {
     "php": "^7.2 || ^8.0",
     "ext-simplexml": "*",
-    "phpstan/phpstan": "^1.11"
+    "phpstan/phpstan": "^1.11",
+    "symfony/contracts": "^2.2"
```
Comment (ondrejmirtes) [inline composer.json]: The extension doesn't really require `symfony/contracts`, please remove this.

**Comment by raalderink** [comment]
I have a couple issues here which I'm having a hard time getting resolved, as they fail on some PHP versions and succeed on others. I've baselined them, but unfortunately `reportUnmatchedIgnoredErrors` is on, which means it will still fail on some and succeed on others.

This following error is actually also baselined in `phpstan-src`, in [enum-adapter-errors.neon](https://github.com/phpstan/phpstan-src/blob/f89852f8c020e7ddc76538f3e6dea510a7271dfd/build/enum-adapter-errors.neon). Due to that, I assume it's a known issue for PHP 7.2 and 7.3, but due to the `reportUnmatchedIgnoredErrors`, I can't really baseline this in this project. How would you want me to solve this issue?
```
-
	message: "#^Call to method getMethods\\(\\) on an unknown class PHPStan\\\\BetterReflection\\\\Reflection\\\\Adapter\\\\ReflectionEnum\\.$#"
	count: 1
	path: src/Symfony/RequiredAutowiringExtension.php
```

Secondly, on the lowest dependency builds, I get the following errors. On the highest dependencies, this builds correctly. What I've done is removed the `class_exists` like you asked, and instead used the `class-string` as string itself. I am having a lot of trouble in finding where this issue lies, and am uncertain of a solution. This is why I had the `class_exists` in the first place, and later tried it by including the actual `symfony-contracts` package that contains it (it makes sense that you don't want that to be included though). Could you please advise on how to approach this and be able to solve this issue?
```
-
	message: "#^Parameter \\#1 \\$name of method PHPStan\\\\BetterReflection\\\\Reflection\\\\Adapter\\\\ReflectionProperty\\:\\:getAttributes\\(\\) expects class\\-string\\|null, string given\\.$#"
	count: 1
	path: src/Symfony/RequiredAutowiringExtension.php

-
	message: "#^Parameter \\#1 \\$name of method PHPStan\\\\BetterReflection\\\\Reflection\\\\Adapter\\\\ReflectionMethod\\:\\:getAttributes\\(\\) expects class\\-string\\|null, string given\\.$#"
	count: 1
	path: src/Symfony/RequiredAutowiringExtension.php
```

These are the only things holding this PR back, so hopefully you can help me out by pointing me towards a solution here. Thanks @ondrejmirtes!

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Solved it, mainly with https://github.com/phpstan/phpstan-src/commit/2b9af363a3d2a9620c4b9bfb65e81813b69623d0, but also locking contracts package version to a newer one :)

**Comment by ondrejmirtes** [comment]
Thank you.
